File: frontend/src/types/contaRecorrente.ts
/**
 * Espelha 1:1 `app/schemas/conta_recorrente.py` — conferido por leitura
 * direta do backend (expansão de Contas Recorrentes, 2026-07-20, ver
 * docs/analise-arquitetural-conta-recorrente-expansao.md).
 *
 * `dia_vencimento` só se aplica a frequências baseadas em meses (MENSAL,
 * BIMESTRAL, TRIMESTRAL, SEMESTRAL, ANUAL) — nulo obrigatório nas baseadas
 * em dias (DIARIA/SEMANAL/QUINZENAL, âncora = `data_inicio`). O backend
 * valida por família; o formulário só condiciona a exibição do campo.
 *
 * `proxima_execucao` é o cursor materializado do backend — a UI mostra
 * "próxima ocorrência" direto daqui, sem recalcular nada no cliente.
 * `status`/`proxima_execucao` nunca são editáveis via PATCH: transições
 * são ações próprias (pausar/reativar/encerrar).
 */
import type { FrequenciaRecorrencia, StatusRecorrencia, TipoTransacao } from "./enums";

export type { FrequenciaRecorrencia, StatusRecorrencia };

export interface ContaRecorrenteRead {
  id: number;
  descricao: string;
  valor: string;
  tipo: TipoTransacao;
  frequencia: FrequenciaRecorrencia;
  dia_vencimento: number | null;
  status: StatusRecorrencia;
  proxima_execucao: string;

  categoria_id: number | null;
  conta_id: number | null;
  cartao_id: number | null;

  data_inicio: string;
  data_fim: string | null;
}

export interface ContaRecorrenteCreate {
  descricao: string;
  valor: string;
  tipo: TipoTransacao;
  frequencia: FrequenciaRecorrencia;
  dia_vencimento?: number | null;

  categoria_id?: number | null;
  conta_id?: number | null;
  cartao_id?: number | null;

  data_inicio: string;
  data_fim?: string | null;
}

/** Semântica de PATCH — campo omitido permanece intocado. */
export interface ContaRecorrenteUpdate {
  descricao?: string;
  valor?: string;
  tipo?: TipoTransacao;
  frequencia?: FrequenciaRecorrencia;
  dia_vencimento?: number | null;

  categoria_id?: number | null;
  conta_id?: number | null;
  cartao_id?: number | null;

  data_inicio?: string;
  data_fim?: string | null;
}

/** Resposta de `POST /contas-recorrentes/sincronizar` — `geradas > 0`
 * decide se os caches são invalidados (evita refetch em cascata a cada
 * login sem novidade). */
export interface SincronizacaoRecorrentesResult {
  geradas: number;
  encerradas: number;
}

File: backend/app/schemas/conta_recorrente.py
"""Schemas de ContaRecorrente: payloads de entrada e saída de
`app/api/routes/conta_recorrente.py`.

Diferente de Fatura/Parcelamento/Transferência, ContaRecorrente TEM um
`ContaRecorrenteUpdate` (PATCH) para os campos do template - editar o
template só afeta ocorrências FUTURAS (cada ocorrência já gerada é uma
Transacao independente, que nunca volta a ler o template depois de criada).

Expansão 2026-07-20 (docs/analise-arquitetural-conta-recorrente-expansao.md):
`dia_vencimento` virou opcional (só se aplica a frequências baseadas em
meses - a validação por família mora no Service, mesmo raciocínio de
sempre); `Read` expõe `status` (ATIVA/PAUSADA/ENCERRADA, substitui `ativo`)
e `proxima_execucao` (o cursor materializado - o frontend mostra "próxima
ocorrência" direto daqui, sem recalcular nada no cliente).

Validação estrutural (`conta_id` XOR `cartao_id`, dia_vencimento ×
frequência) não é feita aqui via `model_validator` - mora inteiramente em
`ContaRecorrenteService`, mesmo raciocínio já documentado em
`app/schemas/transacao.py`.
"""
from datetime import date
from decimal import Decimal

from pydantic import BaseModel, Field

from app.models.enums import FrequenciaRecorrencia, StatusRecorrencia, TipoTransacao
from app.schemas.base import OrmBaseModel


class ContaRecorrenteCreate(BaseModel):
    descricao: str = Field(min_length=1, max_length=200)
    valor: Decimal = Field(gt=0)
    tipo: TipoTransacao
    frequencia: FrequenciaRecorrencia = FrequenciaRecorrencia.MENSAL
    # Obrigatório para frequências baseadas em meses, proibido nas
    # baseadas em dias - validado no Service (erro claro nos 2 sentidos).
    dia_vencimento: int | None = Field(default=None, ge=1, le=31)

    categoria_id: int | None = None
    conta_id: int | None = None
    cartao_id: int | None = None

    data_inicio: date
    data_fim: date | None = None


class ContaRecorrenteUpdate(BaseModel):
    """Todos os campos opcionais - semântica de PATCH. Nunca retroage sobre
    ocorrências já geradas. `status`/`proxima_execucao` NUNCA são editáveis
    por aqui - transições de estado são ações de negócio explícitas
    (pausar/reativar/encerrar), e o cursor é 100% gerenciado pelo Service."""

    descricao: str | None = Field(default=None, min_length=1, max_length=200)
    valor: Decimal | None = Field(default=None, gt=0)
    tipo: TipoTransacao | None = None
    frequencia: FrequenciaRecorrencia | None = None
    dia_vencimento: int | None = Field(default=None, ge=1, le=31)

    categoria_id: int | None = None
    conta_id: int | None = None
    cartao_id: int | None = None

    data_inicio: date | None = None
    data_fim: date | None = None


class SincronizacaoRecorrentesResult(BaseModel):
    """Resposta de `POST /contas-recorrentes/sincronizar` - o frontend usa
    `geradas > 0` para decidir se invalida os caches (evita refetch em
    cascata a cada login sem novidade)."""

    geradas: int
    encerradas: int


class ContaRecorrenteRead(OrmBaseModel):
    id: int
    descricao: str
    valor: Decimal
    tipo: TipoTransacao
    frequencia: FrequenciaRecorrencia
    dia_vencimento: int | None
    status: StatusRecorrencia
    proxima_execucao: date

    categoria_id: int | None
    conta_id: int | None
    cartao_id: int | None

    data_inicio: date
    data_fim: date | None
